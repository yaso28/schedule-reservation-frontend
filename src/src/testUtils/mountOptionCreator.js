import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
