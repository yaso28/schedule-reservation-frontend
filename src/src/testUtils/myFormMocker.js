import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
