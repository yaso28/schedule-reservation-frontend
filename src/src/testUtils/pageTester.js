import { mount } from '@vue/test-utils'
import MyPage from '@/components/containers/MyPage'

class PageTester {
  async mount (pageComponent, route, mountOption) {
    await mountOption.router.push(route)
    const pageWrapper = mount(MyPage, mountOption)
    return pageWrapper.findComponent(pageComponent)
  }

  assertTitleTag () {
    expect(document.title).toMatchSnapshot()
  }

  assertMetaTags () {
    expect(document.querySelectorAll('meta[data-vue-meta]')).toMatchSnapshot()
  }
}

export default new PageTester()
