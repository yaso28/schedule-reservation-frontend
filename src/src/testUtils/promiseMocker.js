import _ from 'lodash'

const createSetTimeoutPromise = getResult => new Promise((resolve, reject) => {
  setTimeout(() => getResult(resolve, reject), 1000)
})
const createSwitchPromise = (value, error, switchObj) => createSetTimeoutPromise(
  (resolve, reject) => switchObj.isSuccess ? resolve(value) : reject(error)
)
const createResolvedPromise = value => createSetTimeoutPromise(resolve => resolve(value))
const createRejectedPromise = error => createSetTimeoutPromise((resolve, reject) => reject(error))

const createMock = createPromise => jest.fn().mockImplementationOnce(createPromise)
const createMockForCases = (caseList, createPromiseForCurrentCase) => jest.fn((...args) => {
  for (let i = 0; i < caseList.length; i++) {
    const currentCase = caseList[i]
    if (_.isEqual(args, currentCase.args)) {
      return createPromiseForCurrentCase(currentCase)
    }
  }
  throw new Error('createMockForCases: no args matched')
})

class PromiseMocker {
  init () {
    jest.useFakeTimers()
  }

  clear () {
    jest.clearAllTimers()
  }

  flushPendings () {
    jest.runOnlyPendingTimers()
  }

  createSwitch () {
    return { isSuccess: true }
  }

  setSwitch (switchObj, isSuccess) {
    switchObj.isSuccess = isSuccess
  }

  mock (value, error, switchObj) {
    return createMock(() => createSwitchPromise(value, error, switchObj))
  }

  mockResolved (value) {
    return createMock(() => createResolvedPromise(value))
  }

  mockRejected (error) {
    return createMock(() => createRejectedPromise(error))
  }

  mockForCases (caseList) {
    return createMockForCases(caseList, currentCase => createSwitchPromise(currentCase.value, currentCase.error, currentCase.switch))
  }

  mockResolvedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createResolvedPromise(currentCase.value))
  }

  mockRejectedForCases (caseList) {
    return createMockForCases(caseList, currentCase => createRejectedPromise(currentCase.error))
  }
}

export default new PromiseMocker()
