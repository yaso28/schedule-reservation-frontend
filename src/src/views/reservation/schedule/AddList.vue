<template>
  <div>
    <slot name="h1" />

    <CForm
      v-if="loaded"
      @submit.prevent="save"
    >
      <CCard>
        <CCardBody>
          <MyArray
            :label="$t('schedule.ymd')"
            :form="form"
            field="ymd_list"
          >
            <template #default="{innerForm, innerField}">
              <MyDateNoWrapper
                :field="innerField"
                :form="innerForm"
                :show-clear="false"
              />
            </template>
            <template #listItem="{item}">
              <MyDateSpan :value="item" />
            </template>
          </MyArray>
          <MyTime
            :label="$t('schedule.begins_at')"
            :form="form"
            field="begins_at"
          />
          <MyTime
            :label="$t('schedule.ends_at')"
            :form="form"
            field="ends_at"
          />
          <MySelect
            :label="$t('schedule_place.model')"
            :form="form"
            field="schedule_place_id"
            :options="placeList"
          />
          <MySelect
            :label="$t('schedule_usage.model')"
            :form="form"
            field="schedule_usage_id"
            :options="usageList"
          />
          <MySelect
            :label="$t('schedule_timetable.model')"
            :form="form"
            field="schedule_timetable_id"
            :options="timetableList"
          />
          <MySelect
            :label="$t('reservation_status.model')"
            :form="form"
            field="reservation_status_id"
            :options="reservationStatusList"
          />
          <MySelect
            :label="$t('schedule_status.model')"
            :form="form"
            field="schedule_status_id"
            :options="scheduleStatusList"
          />
        </CCardBody>
        <CCardFooter>
          <CButton
            class="btn-reset"
            color="secondary"
            @click="form.reset()"
          >
            {{ $t("action.reset") }}
          </CButton>
          <CButton
            type="submit"
            color="primary"
          >
            {{ $t('action.save') }}
          </CButton>
        </CCardFooter>
      </CCard>
    </CForm>
  </div>
</template>

<script>
export default {
  name: 'ReservationScheduleAddList',
  components: {
    MyArray: require('@/components/inputs/MyArray').default,
    MyDateNoWrapper: require('@/components/inputs/MyDateNoWrapper').default,
    MyDateSpan: require('@/components/MyDateSpan').default,
    MyTime: require('@/components/inputs/MyTime').default,
    MySelect: require('@/components/inputs/MySelect').default
  },
  inheritAttrs: false,
  inject: ['apiService'],
  data () {
    return {
      form: null,
      placeList: null,
      usageList: null,
      timetableList: null,
      reservationStatusList: null,
      scheduleStatusList: null
    }
  },
  computed: {
    loaded () {
      return !!(this.form && this.placeList && this.usageList && this.timetableList && this.reservationStatusList && this.scheduleStatusList)
    }
  },
  mounted () {
    this.load()
  },
  methods: {
    async load () {
      try {
        this.$store.commit('processing/beginGet')
        ;[this.placeList, this.usageList, this.timetableList, this.reservationStatusList, this.scheduleStatusList] = await Promise.all([
          this.apiService.getSchedulePlaceList(),
          this.apiService.getScheduleUsageList(),
          this.apiService.getScheduleTimetableList(),
          this.apiService.getReservationStatusList(),
          this.apiService.getScheduleStatusList()
        ])
        this.form = this.apiService.makeForm({
          ymd_list: [],
          begins_at: '',
          ends_at: '',
          schedule_place_id: null,
          schedule_usage_id: null,
          schedule_timetable_id: null,
          reservation_status_id: null,
          schedule_status_id: null
        })
      } catch (error) {
        if (!this.apiService.isHandledError(error)) {
          throw error
        }
      } finally {
        this.$store.commit('processing/endGet')
      }
    },
    async save () {
      try {
        this.$store.commit('processing/beginPost')
        this.$store.commit('alert/clear')
        await this.apiService.addScheduleList(this.form)
        this.$store.dispatch('alert/saveSuccessNoRedirect')
        this.form.reset()
      } catch (error) {
        if (!this.apiService.isHandledError(error)) {
          throw error
        }
      } finally {
        this.$store.commit('processing/endPost')
      }
    }
  }
}
</script>
