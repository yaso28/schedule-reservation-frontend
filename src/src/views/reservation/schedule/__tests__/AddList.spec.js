import AddList from '../AddList'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import pageTester from '@/testUtils/pageTester'
import promiseMocker from '@/testUtils/promiseMocker'
import storeModules from '@/store/modules'
import myFormMocker from '@/testUtils/myFormMocker'

jest.mock('@/store/modules')
jest.mock('@/models/MyForm')

const placeList = require('@/testUtils/dummyApiData/schedule-place/list.json')
const usageList = require('@/testUtils/dummyApiData/schedule-usage/list.json')
const timetableList = require('@/testUtils/dummyApiData/schedule-timetable/list.json')
const scheduleStatusList = require('@/testUtils/dummyApiData/schedule-status/list.json')
const reservationStatusList = require('@/testUtils/dummyApiData/reservation-status/list.json')
const makeFormArgs = {
  ymd_list: [],
  begins_at: '',
  ends_at: '',
  schedule_place_id: null,
  schedule_usage_id: null,
  schedule_timetable_id: null,
  reservation_status_id: null,
  schedule_status_id: null
}

describe('AddList', () => {
  let mountOption
  let apiService
  beforeEach(() => {
    promiseMocker.init()
    mountOption = mountOptionCreator.create(storeModules)
    apiService = mountOption.provide.apiService
  })

  afterEach(() => {
    promiseMocker.clear()
    jest.resetAllMocks()
  })

  describe('マウント', () => {
    let wrapper
    let loadError
    let loadPlaceListSwitch
    let loadUsageListSwitch
    let loadTimetableListSwitch
    let loadReservationStatusListSwitch
    let loadScheduleStatusListSwitch
    beforeEach(async () => {
      loadError = new Error('load')
      loadPlaceListSwitch = promiseMocker.createSwitch()
      loadUsageListSwitch = promiseMocker.createSwitch()
      loadTimetableListSwitch = promiseMocker.createSwitch()
      loadReservationStatusListSwitch = promiseMocker.createSwitch()
      loadScheduleStatusListSwitch = promiseMocker.createSwitch()
      apiService.getSchedulePlaceList = promiseMocker.mock(placeList, loadError, loadPlaceListSwitch)
      apiService.getScheduleUsageList = promiseMocker.mock(usageList, loadError, loadUsageListSwitch)
      apiService.getScheduleTimetableList = promiseMocker.mock(timetableList, loadError, loadTimetableListSwitch)
      apiService.getReservationStatusList = promiseMocker.mock(reservationStatusList, loadError, loadReservationStatusListSwitch)
      apiService.getScheduleStatusList = promiseMocker.mock(scheduleStatusList, loadError, loadScheduleStatusListSwitch)

      wrapper = await pageTester.mount(AddList, { name: 'reservation.schedule.add-list' }, mountOption)
    })

    test('titleタグ', () => {
      pageTester.assertTitleTag()
    })

    test('metaタグ', () => {
      pageTester.assertMetaTags()
    })

    test('Processingカバー表示開始', () => {
      expect(storeModules.processing.mutations.beginGet).toHaveBeenCalled()
    })

    test.each([
      ['場所', apiService => apiService.getSchedulePlaceList],
      ['用途', apiService => apiService.getScheduleUsageList],
      ['時間割', apiService => apiService.getScheduleTimetableList],
      ['予約ステータス', apiService => apiService.getReservationStatusList],
      ['公開ステータス', apiService => apiService.getScheduleStatusList]
    ])('%sマスタ取得処理を呼出', (masterName, getApiMethodToAssert) => {
      expect(getApiMethodToAssert(apiService)).toHaveBeenCalledWith()
    })

    test('Processingカバー表示中', () => {
      expect(
        storeModules.processing.mutations.endGet
      ).not.toHaveBeenCalled()
    })

    describe.each([
      ['場所マスタ'],
      ['用途マスタ'],
      ['時間割マスタ'],
      ['予約ステータスマスタ'],
      ['公開ステータスマスタ']
    ])('%s取得：エラー', (dataName) => {
      beforeEach(() => {
        const loadSwitch =
          dataName === '場所マスタ' ? loadPlaceListSwitch
            : dataName === '用途マスタ' ? loadUsageListSwitch
              : dataName === '時間割マスタ' ? loadTimetableListSwitch
                : dataName === '予約ステータスマスタ' ? loadReservationStatusListSwitch
                  : dataName === '公開ステータスマスタ' ? loadScheduleStatusListSwitch
                    : undefined
        promiseMocker.setSwitch(loadSwitch, false)
        apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
        promiseMocker.flushPendings()
      })

      test('エラー処理', () => {
        expect(apiService.isHandledError).toHaveBeenCalledWith(loadError)
      })

      test('Processingカバー表示終了', () => {
        expect(storeModules.processing.mutations.endGet).toHaveBeenCalled()
      })
    })

    describe('データ取得：成功', () => {
      let form
      beforeEach(() => {
        form = myFormMocker.createMock(makeFormArgs)
        apiService.makeForm = jest.fn().mockReturnValueOnce(form)
        promiseMocker.flushPendings()
      })

      test('フォーム作成', () => {
        expect(apiService.makeForm).toHaveBeenCalledWith(makeFormArgs)
      })

      test('ページ全体表示', () => {
        expect(wrapper.element).toMatchSnapshot()
      })

      test('Processingカバー表示終了', () => {
        expect(storeModules.processing.mutations.endGet).toHaveBeenCalled()
      })

      describe('リセットボタンクリック', () => {
        beforeEach(() => {
          wrapper.find('button.btn-reset').trigger('click')
        })

        test('フォームリセット', () => {
          expect(form.reset).toHaveBeenCalledWith()
        })
      })

      describe('フォーム送信', () => {
        let saveError
        let saveSwitch
        beforeEach(() => {
          saveError = new Error('save')
          saveSwitch = promiseMocker.createSwitch()
          apiService.addScheduleList = promiseMocker.mock(null, saveError, saveSwitch)
          wrapper.find('form').trigger('submit')
        })

        test('Processingカバー表示開始', () => {
          expect(storeModules.processing.mutations.beginPost).toHaveBeenCalled()
        })

        test('アラートクリア', () => {
          expect(storeModules.alert.mutations.clear).toHaveBeenCalled()
        })

        test('保存処理を呼出', () => {
          expect(apiService.addScheduleList).toHaveBeenCalledWith(form)
        })

        test('Processingカバー表示中', () => {
          expect(storeModules.processing.mutations.endPost).not.toHaveBeenCalled()
        })

        describe('保存：エラー', () => {
          beforeEach(() => {
            promiseMocker.setSwitch(saveSwitch, false)
            apiService.isHandledError = jest.fn().mockReturnValueOnce(true)
            promiseMocker.flushPendings()
          })

          test('エラー処理', () => {
            expect(apiService.isHandledError).toHaveBeenCalledWith(saveError)
          })

          test('Processingカバー表示終了', () => {
            expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
          })

          test('成功アラートは呼び出さない', () => {
            expect(storeModules.alert.actions.saveSuccessNoRedirect).not.toHaveBeenCalled()
          })

          test('フォームリセットは呼び出さない', () => {
            expect(form.reset).not.toHaveBeenCalled()
          })
        })

        describe('保存：成功', () => {
          beforeEach(() => {
            promiseMocker.setSwitch(saveSwitch, true)
            promiseMocker.flushPendings()
          })

          test('成功アラート', () => {
            expect(storeModules.alert.actions.saveSuccessNoRedirect).toHaveBeenCalled()
          })

          test('フォームリセット', () => {
            expect(form.reset).toHaveBeenCalledWith()
          })

          test('Processingカバー表示終了', () => {
            expect(storeModules.processing.mutations.endPost).toHaveBeenCalled()
          })
        })
      })
    })
  })
})
