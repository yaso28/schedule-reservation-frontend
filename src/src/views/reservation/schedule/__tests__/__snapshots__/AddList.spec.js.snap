// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddList マウント metaタグ 1`] = `
NodeList [
  <meta
    content="noindex"
    data-vmid="robots"
    data-vue-meta="2"
    name="robots"
  />,
]
`;

exports[`AddList マウント titleタグ 1`] = `"やそスポーツクラブ | 練習予定追加"`;

exports[`AddList マウント データ取得：成功 ページ全体表示 1`] = `
<div>
  <h1>
    練習予定追加
  </h1>
   
  <form
    class=""
  >
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <div
          class="form-row form-group"
        >
          <label
            class="col-form-label col-sm-3"
            for="ymd_list"
          >
            
    日付
  
          </label>
           
          <div
            class="col-sm-9"
          >
            <div
              class="my-input-wrapper"
            >
              <div>
                <!---->
                 
                <ul
                  class="list-group my-array-items"
                  role="list-items"
                />
              </div>
            </div>
             
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="begins_at"
          >
             開始時刻 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <input
              class="form-control"
              id="begins_at"
              step="900"
              type="time"
            />
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="ends_at"
          >
             終了時刻 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <input
              class="form-control"
              id="ends_at"
              step="900"
              type="time"
            />
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="schedule_place_id"
          >
             場所 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <select
              class="form-control"
              id="schedule_place_id"
            >
              <!---->
              <option
                data-key="0"
                value=""
              >
                  
              </option>
              <option
                data-key="1"
                value="1"
              >
                 体育館＠公民館別棟 
              </option>
              <option
                data-key="2"
                value="2"
              >
                 多目的ホール＠公民館1F 
              </option>
              <option
                data-key="3"
                value="3"
              >
                 大会議室＠公民館2F 
              </option>
            </select>
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="schedule_usage_id"
          >
             用途 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <select
              class="form-control"
              id="schedule_usage_id"
            >
              <!---->
              <option
                data-key="0"
                value=""
              >
                  
              </option>
              <option
                data-key="1"
                value="1"
              >
                 通常練習 
              </option>
              <option
                data-key="2"
                value="2"
              >
                 体験会 
              </option>
              <option
                data-key="3"
                value="3"
              >
                 会議 
              </option>
            </select>
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="schedule_timetable_id"
          >
             時間割 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <select
              class="form-control"
              id="schedule_timetable_id"
            >
              <!---->
              <option
                data-key="0"
                value=""
              >
                  
              </option>
              <option
                data-key="1"
                value="1"
              >
                 午前 
              </option>
              <option
                data-key="2"
                value="2"
              >
                 午後 
              </option>
              <option
                data-key="3"
                value="3"
              >
                 一日 
              </option>
            </select>
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="reservation_status_id"
          >
             予約ステータス 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <select
              class="form-control"
              id="reservation_status_id"
            >
              <!---->
              <option
                data-key="0"
                value=""
              >
                  
              </option>
              <option
                data-key="1"
                value="1"
              >
                 未予約 
              </option>
              <option
                data-key="2"
                value="2"
              >
                 電話予約済 
              </option>
              <option
                data-key="3"
                value="3"
              >
                 書類提出済 
              </option>
              <option
                data-key="4"
                value="4"
              >
                 書類控取得 
              </option>
              <option
                data-key="5"
                value="5"
              >
                 料金支払済 
              </option>
              <option
                data-key="6"
                value="6"
              >
                 会計精算済 
              </option>
              <option
                data-key="7"
                value="7"
              >
                 キャンセル 
              </option>
            </select>
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
         
        <div
          class="form-group form-row"
          role="group"
        >
          <label
            class="col-form-label col-sm-3"
            for="schedule_status_id"
          >
             公開ステータス 
          </label>
          <div
            class="col-sm-9"
          >
            <!---->
            <select
              class="form-control"
              id="schedule_status_id"
            >
              <!---->
              <option
                data-key="0"
                value=""
              >
                  
              </option>
              <option
                data-key="1"
                value="1"
              >
                 未定 
              </option>
              <option
                data-key="2"
                value="2"
              >
                 確定 
              </option>
              <option
                data-key="3"
                value="3"
              >
                 中止 
              </option>
              <option
                data-key="4"
                value="4"
              >
                 非公開 
              </option>
            </select>
            <!---->
            <!---->
            <!---->
            <!---->
          </div>
        </div>
      </div>
       
      <footer
        class="card-footer"
      >
        <button
          class="btn btn-reset btn-secondary"
          type="button"
        >
          
          リセット
        
        </button>
         
        <button
          class="btn btn-primary"
          type="submit"
        >
          
          保存
        
        </button>
      </footer>
    </div>
  </form>
</div>
`;
